fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java

```diff
@@ -70,19 +72,44 @@ public Expression visit(Expression expr, Void context) {
 
             @Override
             public Expression visitComparisonPredicate(ComparisonPredicate cp, Void context) {
-                if (cp.left().isSlot() && cp.right().isConstant()) {
-                    return replaceSlot(cp);
-                } else if (cp.left().isConstant() && cp.right().isSlot()) {
-                    return replaceSlot(cp);
+                if (isExpressionSlotCoveredByCast(cp.left()) && (cp.right().isConstant())) {
+                    return replaceSlot(cp, getDatatypeCoveredByCast(cp.left()));
+                } else if (isExpressionSlotCoveredByCast(cp.right()) && cp.left().isConstant()) {
+                    return replaceSlot(cp, getDatatypeCoveredByCast(cp.right()));
                 }
                 return super.visit(cp, context);
             }
 
-            private Expression replaceSlot(Expression expr) {
+            private boolean isTwoExpressionEqualWithCast(Expression left, Expression right) {
+                return getSlotCoveredByCast(left).equals(getSlotCoveredByCast(right));
+            }
+
+            private boolean canCompareType(DataType dataType) {
```

Comment:
if u want get all integer type , u could use `instance of IntegralType`

Reply:
done